examples/Example17/src/ParticleGun.cc

```diff
@@ -0,0 +1,48 @@
+// SPDX-FileCopyrightText: 2023 CERN
+// SPDX-License-Identifier: Apache-2.0
+//
+#include "ParticleGun.hh"
+#include "G4ParticleGun.hh"
+#include "G4PhysicalConstants.hh"
+#include "G4PrimaryParticle.hh"
+#include "G4Event.hh"
+#include "Randomize.hh"
+#include "G4SystemOfUnits.hh"
+
+void ParticleGun::GenerateRandomPrimaryVertex(G4Event *aEvent, G4double aMinPhi, G4double aMaxPhi, G4double aMinTheta,
+                                              G4double aMaxTheta, std::vector<G4ParticleDefinition *> *aParticleList)
+{
+  // Choose a particle from the list
+  G4ParticleDefinition *aParticleDefinition = (*aParticleList)[(int)(G4UniformRand() * aParticleList->size())];
```

Comment:
Here you will need to take into account the fraction for each particle type. The typical algorithm does:
```
  auto rand = G4UniformRand();
  size_t ipart     = 0;
  float weight = fractionsList[ipart];
  while (rand > weight && ipart < aParticleList->size())
    weight +=  fractionsList[++ipart];
  assert(ipart < aParticleList->size());
```